pkg/controller/v1beta1/inferenceservice/components/predictor.go

```diff
@@ -241,6 +247,129 @@ func (p *Predictor) Reconcile(isvc *v1beta1.InferenceService) (ctrl.Result, erro
 		}
 	}
 
+	if isvc.Spec.Predictor.WorkerSpec != nil {
+		mergedWorkerServingRuntimeSpec, err := isvcutils.MergeServingRuntimePodSpec(&sRuntime.WorkerSpec.ServingRuntimePodSpec, &isvc.Spec.Predictor.WorkerSpec.PodSpec)
+		if err != nil {
+			isvc.Status.UpdateModelTransitionStatus(v1beta1.InvalidSpec, &v1beta1.FailureInfo{
+				Reason:  v1beta1.InvalidPredictorSpec,
+				Message: "Failed to consolidate serving runtime WorkerSpec ServingRuntimePodSpec",
+			})
+			return ctrl.Result{}, errors.Wrapf(err, "failed to consolidate serving runtime WorkerSpec ServingRuntimePodSpec")
+		}
+
+		if sRuntime.WorkerSpec == nil {
+			sRuntime.WorkerSpec = &v1alpha1.WorkerSpec{}
+		}
+
+		sRuntime.WorkerSpec.ServingRuntimePodSpec = *mergedWorkerServingRuntimeSpec
+	}
+
+	if sRuntime.WorkerSpec != nil {
+		multiNodeEnabled = true
+		// Check if workerSpec does not have containers information, it should return errors
+		if len(sRuntime.WorkerSpec.Containers) == 0 {
+			isvc.Status.UpdateModelTransitionStatus(v1beta1.InvalidSpec, &v1beta1.FailureInfo{
+				Reason:  v1beta1.InvalidPredictorSpec,
+				Message: "No workerSpec container configuration found in selected serving runtime",
+			})
+			return ctrl.Result{}, errors.New("no workerSpec container configuration found in selected serving runtime")
+		}
+		workerContainerIdx := -1
+		for i := range sRuntime.Containers {
+			if sRuntime.WorkerSpec.Containers[i].Name == constants.WorkerContainerName {
+				workerContainerIdx = i
+				break
+			}
+		}
+		if workerContainerIdx == -1 {
+			return ctrl.Result{}, fmt.Errorf("failed to find %s in ServingRuntime workerSpec containers", constants.WorkerContainerName)
+		}
+
+		var workerSpecContainer *v1.Container
+		if isvc.Spec.Predictor.WorkerSpec != nil && isvc.Spec.Predictor.WorkerSpec.Containers != nil {
+			workerSpecContainer = &isvc.Spec.Predictor.WorkerSpec.Containers[0]
+		} else {
+			workerSpecContainer = &v1.Container{}
+		}
+
+		workerContainer, err := isvcutils.MergeRuntimeContainers(&sRuntime.WorkerSpec.Containers[workerContainerIdx], workerSpecContainer)
+		if err != nil {
+			isvc.Status.UpdateModelTransitionStatus(v1beta1.InvalidSpec, &v1beta1.FailureInfo{
+				Reason:  v1beta1.InvalidPredictorSpec,
+				Message: "Failed to get runtime container",
+			})
+			return ctrl.Result{}, errors.Wrapf(err, "failed to get runtime container")
+		}
+		// This is for using isvcutils.MergePodSpec for workerSpec
+		if isvc.Spec.Predictor.WorkerSpec == nil {
+			isvc.Spec.Predictor.WorkerSpec = &v1beta1.WorkerSpec{
+				PodSpec: v1beta1.PodSpec{},
+			}
+		}
+		mergedWorkerPodSpec, err := isvcutils.MergePodSpec(&sRuntime.WorkerSpec.ServingRuntimePodSpec, &isvc.Spec.Predictor.WorkerSpec.PodSpec)
```

Comment:
In line 251 you did:
```go
isvcutils.MergeServingRuntimePodSpec(&sRuntime.WorkerSpec.ServingRuntimePodSpec, &isvc.Spec.Predictor.WorkerSpec.PodSpec)
```

and in line 264 you did:
```go
sRuntime.WorkerSpec.ServingRuntimePodSpec = *mergedWorkerServingRuntimeSpec
```

To my eye, this merge is not needed. Or what's the difference?

Reply:
The purpose of this is to merge the content of the WorkerSpec into the ServingRuntime when the customer specifies it in the InferenceService. So if user set podContiner for workerspec in the isvc, it will be reflected to servingruntime.